Updated the main.tf of gcc to add the ServiceAgentV2Ext role

kiranjotmanjotra-ssk opened the pull request:
- Updated the main.tf of gcc to add the ServiceAgentV2Ext role to the `@cloudcomposer-accounts.iam.gserviceaccount.com` service account.

dmedora commented:
/gcbrun

dmedora commented:
Hi @kiranjotmanjotra-ssk, thanks for the PR!

I wanted to clarify what the goal is here - the title mentions adding `roles/ServiceAgentV2Ext` to the Composer service agent, but I believe that's what's already being done by [that block](https://github.com/GoogleCloudPlatform/data-migration-tool/blob/main/terraform/translation/gcc/main.tf#L51) - `local.cloud_composer_sa` is the service agent.

It looks like this change is adding `depends_on = [google_service_account.composer_service_account]`. Could you help me understand what issue you faced and how this resolves it? `composer_service_account` refers to the user-provided service account that will be bound to the Composer env, which is separate and independent from the Composer Service Agent.

kiranjotmanjotra-ssk commented:
Hi @dmedora,

I ran into an issue while building a Cloud Composer environment. There was an error message indicating: "Cloud Composer Service Agent service account doesn't exist yet, so the `Cloud Composer v2 API Service Agent Extension` role couldn't be assigned to it."

**Explanation:** Cloud Composer automatically creates a service agent and its corresponding service account. However, our current Terraform configuration tries to assign a role (roles/composer.ServiceAgentV2Ext) to this account before it's even created. The current terraform configuration tries to create a Cloud Composer service agent service account before the creation of the cloud composer environment and cloud composer default service account. Hence, it throws an error.

**Solution:** We've adjusted the Terraform code to ensure that the cloud composer environment and its corresponding service account is created first. This ensures a smooth workflow for attaching the desired roles to the Cloud Composer Service Agent Service Account.

**Note:** Previously, we attempted to make the role assignment dependent on the creation of the cloud composer environment itself. This resulted in a new error message about missing permissions for the Cloud composer service agent service account. The current solution avoids that issue by focusing on the service agent account creation order.

dmedora commented:
Thanks for the clear explanation @kiranjotmanjotra-ssk!

Just to clarify, Composer only creates the one Composer Service Agent. The service account being created (`composer_service_account` TF resource) is just a normal user-managed service account. It does need to be created before the Composer env, since we provide it at env creation time (and there is a depends_on for that).

The service agent should be created soon after enabling the Composer API. As you mentioned, it needs to be granted the ServiceAgentV2Ext role before attempting to create the env. I'm not sure why it's working for you to delay the permission assignment until after creating the user-managed service account, but I'm guessing it just gives enough time for the agent's auto-creation. We haven't run into this before - I'm guessing it only happens when you try to deploy immediately after creating a new project with TF as well? 

I think it might be clearer to just use a `sleep` block to wait a minute after API enablement before attempting the IAM binding, since that's what we would be implicitly doing by add `depends_on composer_service_account` given there is no actual dependency between the two.

Raniksingh commented:
/gcbrun

kiranjotmanjotra-ssk commented:
Hi @dmedora,

I've implemented the `sleep` block, in the main.tf file and removed the dependency on `depends_on composer_service_account` and everything seems to be functioning as intended.

Thanks
